### ISSUE TITLE
Keep releases & documentation synchronized

### ISSUE THREAD

**Issue by thilp** [zalando-incubator/Transformer#28]
## Problem

When @lispyclouds tried Transformer, the stable release was 1.0.1 (which is what he got from PyPI).

**However**, some work had _already been merged to master_: in particular, the addition of the `dump` method and the README update that made use of `dump` in the main example.

So the first thing our user tried is `dump`, which did not exist in his release (from PyPI, on 1.0.1) but was documented (in master, on 1.0.2.dev*). He experienced a crash and, had he not the chance (?) to sit close to Transformer's maintainers (who told him to use `pip install --pre` to get 1.0.2.dev), he would probably have concluded that this project was just broken.

**How** should we release new versions and maintain our documentation so that this doesn't happen?

## Ideas

1. Every commit on master is a stable PyPI release. Prerelease development happens on a `dev` branch, from which (resp. into which) every feature branch starts (resp. is merged).
    - Downside: This make us no longer stick to [GitHub Flow](https://guides.github.com/introduction/flow/), which may be confusing. Contributors will likely often try to merge their work into master out of habit.
    - Downside: This solves our problem with respect to documentation _only if we track the documentation with the code_ (as in our example, with the readme). However, we recently started using GitHub's wiki feature because it is more comfortable. Unless we add constraints like branches on the wiki too, our wiki documentation will always reflect the most recent, prerelease state.
1. Same as above but master stays the "dev" branch: A new, protected branch `release` is created instead.
    - Downside: Same as above with respect to documentation in the wiki.
1. When documentation is updated, the next stable version is mentioned. Example: "Changed in v1.0.2: You can use the `dump` method etc.".
    - Downside: This looks verbose and ad-hoc. We should come up with a standardized format to make writing and reading such annotations easier.
1. Every commit on `master` is a stable PyPI release. And that's it: no alternative "prerelease" branch.
    - Downside: All PRs must go through the release process, in particular the full changelog update (links, etc.), which can be painful when multiple PRs are worked on concurrently.
    - Downside: You can only write the documentation (in the wiki) _after_ the PR is merged, otherwise we're not solving this issue. So the documentation is not reviewed and we can even forget to add it, since it happens after the fact. This can be mitigated:
        - Update the documentation before the PR is merged but with annotations like "prerelease: …". The documentation must be updated again after the PR is merged to remove these annotations, and the same kind of problems apply (can be forgotten, etc.). ~No easy or automatic way to check that the current annotations are still relevant.~ We can include the PR number in the annotation for easier tracking.
        - Opening an issue about updating the documentation is a mandatory part of the PR process. Downside: more paperwork. But at least it's tracked properly, and the issue can contain the planned changes so that they can be reviewed before the PR merge.

**Comment by thilp** [comment]
@tortila @bmaher I'd really like to know your thoughts on that topic.

Anyone's opinion is of course welcome!

**Comment by tortila** [comment]
As we discussed face to face, I think it's important to focus on continuous releases, and at the same time somehow find a solution to track changes in documentation in the same manner as we do with the code. I agree that Wiki has its advantages and is more user-friendly than READMEs, which we used previously, but at the same time it forces us to trade off the visibility of changes. 

As an alternative, [GitHub Project Pages](https://help.github.com/en/articles/user-organization-and-project-pages#project-pages-sites) provides a nice mechanism for hosting static documentation which sources come from the same repository as the project itself. I noticed that there are 3 different conventions to use the GitHub Pages, so if we decide to go with this solution, we should think which one suits our use case best.
